### PR TITLE
Reduce async overhead in CodeAnalysis

### PR DESCRIPTION
**Pull request by pentp**
This removes unnecessary async state machines from functions that forward to other async functions.

### CODE REVIEW

File: src/Workspaces/Core/Portable/CodeFixes/FixAllOccurrences/FixAllContext.cs [hunk 1]
```diff
@@ -139,7 +139,7 @@ public partial class FixAllContext
         /// <summary>
         /// Gets all the diagnostics in the given document filtered by <see cref="DiagnosticIds"/>.
         /// </summary>
-        public async Task<ImmutableArray<Diagnostic>> GetDocumentDiagnosticsAsync(Document document)
+        public Task<ImmutableArray<Diagnostic>> GetDocumentDiagnosticsAsync(Document document)
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/CodeFixes/FixAllOccurrences/FixAllContext.cs]: I would not change this one.

File: src/Workspaces/Core/Portable/CodeFixes/FixAllOccurrences/FixAllContext.cs [hunk 2]
```diff
@@ -172,48 +172,48 @@ public partial class FixAllContext
...
         /// </summary>
-        private async Task<ImmutableArray<Diagnostic>> GetProjectDiagnosticsAsync(Project project, bool includeAllDocumentDiagnostics)
+        private Task<ImmutableArray<Diagnostic>> GetProjectDiagnosticsAsync(Project project, bool includeAllDocumentDiagnostics)
         {
             Contract.ThrowIfNull(project);
 
             if (this.Project.Language != project.Language)
             {
-                return ImmutableArray<Diagnostic>.Empty;
+                return Task.FromResult(ImmutableArray<Diagnostic>.Empty);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/CodeFixes/FixAllOccurrences/FixAllContext.cs]: Or this one.

File: src/Workspaces/Core/Portable/CodeFixes/FixAllOccurrences/FixAllContext.cs (hunk 1 above)
Comment (pentp) [reply src/Workspaces/Core/Portable/CodeFixes/FixAllOccurrences/FixAllContext.cs]: OK, I'll revert, but is this technically invalid or just bad for code readability?

**Comment by sharwell** [comment]
:memo: Considering stack traces get a bit harder to decipher each time this approach is used, I would prefer to avoid making the change for any case we cannot prove is on a hot path for which the change will have a measurable difference in some particular user scenario.

**Comment by pentp** [comment]
I reverted changes where important stack trace information could get lost. The other places are either forwarding to similar internal functions or have their stack trace information in lambdas (e.g., `RunServiceAsync` invocations).

File: src/Workspaces/Core/Portable/Shared/Extensions/DocumentExtensions.cs [hunk 3]
```diff
@@ -176,17 +176,17 @@ public static Task<SyntaxTreeIndex> GetSyntaxTreeIndexAsync(this Document docume
         /// Returns the semantic model for this document that may be produced from partial semantics. The semantic model
         /// is only guaranteed to contain the syntax tree for <paramref name="document"/> and nothing else.
         /// </summary>
-        public static async Task<SemanticModel> GetPartialSemanticModelAsync(this Document document, CancellationToken cancellationToken)
+        public static Task<SemanticModel> GetPartialSemanticModelAsync(this Document document, CancellationToken cancellationToken)
         {
-            if (document.Project.TryGetCompilation(out var compilation))
+            if (document.Project.TryGetCompilation(out _))
```
Comment (jasonmalinowski) [inline src/Workspaces/Core/Portable/Shared/Extensions/DocumentExtensions.cs]: Hard to tell if this actually is losing some of the original intent of this function: if we did fetch a compilation, we'd want to make sure the compilation is held alive prior to calling document.GetSemanticModelAsync(). The function probably should still capture compilation, then do:

```
var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
GC.KeepAlive(compilation);
return semanticModel;
```

The lack of GC.KeepAlive here is probably bad always, but discarding the variable _might_ impact JIT and thus make things even worse.

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.cs [hunk 4]
```diff
@@ -876,15 +876,16 @@ public Task<bool> HasSuccessfullyLoadedAsync(SolutionState solution, Cancellatio
...
             private AsyncLazy<VersionStamp> _lazyDependentSemanticVersion;
 
-            public async Task<VersionStamp> GetDependentVersionAsync(SolutionState solution, CancellationToken cancellationToken)
+            public Task<VersionStamp> GetDependentVersionAsync(SolutionState solution, CancellationToken cancellationToken)
             {
                 if (_lazyDependentVersion == null)
                 {
                     // note: solution is captured here, but it will go away once GetValueAsync executes.
-                    Interlocked.CompareExchange(ref _lazyDependentVersion, new AsyncLazy<VersionStamp>(c => ComputeDependentVersionAsync(solution, c), cacheResult: true), null);
+                    var tmp = solution;
```
Comment (jasonmalinowski) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.cs]: Why was this necessary here?

Comment (pentp) [reply src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.cs]: This avoids a closure allocation for the fast path.

File: src/Workspaces/Core/Portable/Shared/Extensions/DocumentExtensions.cs (hunk 3 above)
Comment (pentp) [reply src/Workspaces/Core/Portable/Shared/Extensions/DocumentExtensions.cs]: I verified that there is no IL codegen difference between these 3 variants (async + out var, out var, discard) - all discard the value, so the compilation wasn't kept alive before also. Should I restore the async state machine here and add a GC.KeepAlive instead then?

**Comment by CyrusNajmabadi** [comment]
Also consider that a lot of these codepaths could really benefit from moving to ValueTask.  @sharwell @jasonmalinowski is that on your radar at all?

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.cs (hunk 4 above)
Comment (jasonmalinowski) [reply src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.cs]: Ah, I see. Can you add a comment to that effect?

**Comment by jasonmalinowski** [comment]
@CyrusNajmabadi @sharwell Has begun to look at it. But early stages yet.

**Comment by sharwell** [comment]
@CyrusNajmabadi `ValueTask<T>` will incur an assembly load so we have to make a very compelling performance case. So far the measured improvement is 0.5% (which will grow as other items get fixed).

**Comment by CyrusNajmabadi** [comment]
@sharwell .5% is pretty great.  Seems worth it to me.  Hope that management approves :)

**Comment by pentp** [comment]
I just added two comments and now a test is failing...

**Comment by CyrusNajmabadi** [comment]
> I just added two comments and now a test is failing...

You did change Completion (the recommender service).  So it's possible this had a visible change somewhere.

File: src/Workspaces/Core/Portable/CodeCleanup/Providers/SimpleCodeCleanupProvider.cs [hunk 5]
```diff
@@ -31,13 +31,18 @@ internal class SimpleCodeCleanupProvider : ICodeCleanupProvider
...
             if (_documentDelegatee != null)
             {
-                return await _documentDelegatee(document, spans, cancellationToken).ConfigureAwait(false);
+                return _documentDelegatee(document, spans, cancellationToken);
             }
 
+            return CleanupAsyncCore(document, spans, cancellationToken);
+        }
+
+        private async Task<Document> CleanupAsyncCore(Document document, ImmutableArray<TextSpan> spans, CancellationToken cancellationToken)
```
Comment (sharwell) [inline src/Workspaces/Core/Portable/CodeCleanup/Providers/SimpleCodeCleanupProvider.cs]: ❗️ Asynchronous method is missing `Async` suffix

**Comment by sharwell** [comment]
@pentp I have a somewhat similar change in #23557. This morning I created a new `[PerformanceSensitive]` attribute which allows performance sensitive code to link back to an issue that describes the test process for validating changes. I'm not expecting this pull request to necessarily use that attribute (it hasn't been merged yet), but in preparation for it we should make sure to explicitly document the set of steps to reveal the performance overhead of each of these cases.

File: src/Workspaces/Core/Portable/Shared/Extensions/DocumentExtensions.cs [hunk 6]
```diff
@@ -181,7 +181,9 @@ public static Task<SyntaxTreeIndex> GetSyntaxTreeIndexAsync(this Document docume
             if (document.Project.TryGetCompilation(out var compilation))
             {
                 // We already have a compilation, so at this point it's fastest to just get a SemanticModel
-                return await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+                var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+                GC.KeepAlive(compilation);
```
Comment (sharwell) [inline src/Workspaces/Core/Portable/Shared/Extensions/DocumentExtensions.cs]: ❗️ This change would need to be in a separate pull request (`GC.KeepAlive` requires additional review).

File: src/Workspaces/Remote/ServiceHub/Services/CodeAnalysisService_AddImport.cs [hunk 7]
```diff
@@ -58,31 +58,25 @@ public SymbolSearchService(CodeAnalysisService codeAnalysisService)
                 this.codeAnalysisService = codeAnalysisService;
             }
 
-            public async Task<IList<PackageWithTypeResult>> FindPackagesWithTypeAsync(
+            public Task<IList<PackageWithTypeResult>> FindPackagesWithTypeAsync(
```
Comment (sharwell) [inline src/Workspaces/Remote/ServiceHub/Services/CodeAnalysisService_AddImport.cs]: :memo: This one makes me uncomfortable (loses the name `FindPackagesWithTypeAsync` in the stack trace, plus this is an OOP call so the overhead of the async state machine is minimal in comparison.

File: src/Workspaces/Remote/ServiceHub/Services/CodeAnalysisService_NavigateTo.cs [hunk 8]
```diff
@@ -43,10 +43,10 @@ internal partial class CodeAnalysisService : IRemoteNavigateToSearchService
 
                     return Convert(result);
                 }
-            }, cancellationToken).ConfigureAwait(false);
+            }, cancellationToken);
         }
 
-        private ImmutableArray<SerializableNavigateToSearchResult> Convert(
+        private IList<SerializableNavigateToSearchResult> Convert(
```
Comment (sharwell) [inline src/Workspaces/Remote/ServiceHub/Services/CodeAnalysisService_NavigateTo.cs]: :bulb: Generally not a fan of changing return types from immutable data to mutable (increases maintenance cost and complexity)

File: src/Workspaces/Core/Portable/FindSymbols/SymbolFinder_FindReferences_Current.cs [hunk 9]
```diff
@@ -51,7 +51,7 @@ public static partial class SymbolFinder
             return engine.FindReferencesAsync(symbolAndProjectId);
         }
 
-        private static async Task<bool> TryFindReferencesInServiceProcessAsync(
+        private static Task<bool> TryFindReferencesInServiceProcessAsync(
```
Comment (sharwell) [inline src/Workspaces/Core/Portable/FindSymbols/SymbolFinder_FindReferences_Current.cs]: ❗️ This one changes the names in the asynchronous stack trace.

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.cs [hunk 10]
```diff
@@ -876,15 +876,16 @@ public Task<bool> HasSuccessfullyLoadedAsync(SolutionState solution, Cancellatio
             private AsyncLazy<VersionStamp> _lazyDependentVersion;
             private AsyncLazy<VersionStamp> _lazyDependentSemanticVersion;
 
-            public async Task<VersionStamp> GetDependentVersionAsync(SolutionState solution, CancellationToken cancellationToken)
+            public Task<VersionStamp> GetDependentVersionAsync(SolutionState solution, CancellationToken cancellationToken)
             {
                 if (_lazyDependentVersion == null)
                 {
+                    var tmp = solution; // temp. local to avoid a closure allocation for the fast path
```
Comment (sharwell) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.cs]: :memo: I'm planning to propose an analyzer to enforce this as a constraint so that should help prevent regressions.

**Comment by sharwell** [comment]
My primary remaining issues are the following:

* When removing `async`, validating the change to exception behavior is challenging. This increases the risk associated with the change.
* Some of the changed methods forward the call to a method with a different name, so the asynchronous stack trace will lose some of its relevance.
* Its difficult to tell which methods are changing "just because they can", and which methods are changing because they are causing and observable/measurable problem.

File: src/Workspaces/Core/Portable/CodeFixes/FixAllOccurrences/FixAllContext.cs [hunk 11]
```diff
@@ -172,28 +172,28 @@ public partial class FixAllContext
...
-            return await GetProjectDiagnosticsAsync(project, includeAllDocumentDiagnostics: false).ConfigureAwait(false);
+            return GetProjectDiagnosticsAsync(project, includeAllDocumentDiagnostics: false);
         }
 
         /// <summary>
         /// Gets all the diagnostics in the given project filtered by <see cref="DiagnosticIds"/>.
         /// This includes both document-level diagnostics for all documents in the given project and project-level diagnostics, i.e. diagnostics with no source location, in the given project. 
         /// </summary>
-        public async Task<ImmutableArray<Diagnostic>> GetAllDiagnosticsAsync(Project project)
+        public Task<ImmutableArray<Diagnostic>> GetAllDiagnosticsAsync(Project project)
```
[the remark this answers is not in the document]
Comment (tmat) [reply src/Workspaces/Core/Portable/CodeFixes/FixAllOccurrences/FixAllContext.cs]: Passing null to project is a programmer error. The new behavior is preferred, imo. It is a change to public API but it only affects code that's already incorrect.

File: src/Workspaces/Core/Portable/CodeFixes/FixAllOccurrences/FixAllContext.cs [hunk 12]
```diff
@@ -172,28 +172,28 @@ public partial class FixAllContext
         /// <summary>
         /// Gets all the project-level diagnostics, i.e. diagnostics with no source location, in the given project filtered by <see cref="DiagnosticIds"/>.
         /// </summary>
-        public async Task<ImmutableArray<Diagnostic>> GetProjectDiagnosticsAsync(Project project)
+        public Task<ImmutableArray<Diagnostic>> GetProjectDiagnosticsAsync(Project project)
```
[the remark this answers is not in the document]
Comment (sharwell) [reply src/Workspaces/Core/Portable/CodeFixes/FixAllOccurrences/FixAllContext.cs]: ➡️ Can leave this.

File: src/Workspaces/Core/Portable/CodeFixes/FixAllOccurrences/FixAllContext.cs (hunk 11 above)
[the remark this answers is not in the document]
Comment (sharwell) [reply src/Workspaces/Core/Portable/CodeFixes/FixAllOccurrences/FixAllContext.cs]: Sure.

➡️ Can leave this.

File: src/Workspaces/Core/Portable/FindSymbols/SymbolFinder_FindLiteralReferences.cs [hunk 13]
```diff
@@ -41,7 +41,7 @@ public static partial class SymbolFinder
             return engine.FindReferencesAsync();
         }
 
-        private static async Task<bool> TryFindLiteralReferencesInServiceProcessAsync(
+        private static Task<bool> TryFindLiteralReferencesInServiceProcessAsync(
```
Comment (sharwell) [inline src/Workspaces/Core/Portable/FindSymbols/SymbolFinder_FindLiteralReferences.cs]: 📝 This one seems like a name change that could be confusing.

File: src/Workspaces/Core/Portable/Workspace/Solution/TextDocumentState.cs [hunk 14]
```diff
@@ -369,15 +369,15 @@ public TextDocumentState UpdateText(TextLoader loader, PreservationMode mode)
                 lazyChecksums: null);
         }
 
-        private async Task<TextAndVersion> GetTextAndVersionAsync(CancellationToken cancellationToken)
+        private Task<TextAndVersion> GetTextAndVersionAsync(CancellationToken cancellationToken)
```
Comment (sharwell) [inline src/Workspaces/Core/Portable/Workspace/Solution/TextDocumentState.cs]: 💭 Not sure about this one. Maybe it's easy to understand in call stacks but maybe not.

File: src/Workspaces/Remote/ServiceHub/Services/CodeAnalysisService_NavigateTo.cs (hunk 8 above)
Comment (pentp) [reply src/Workspaces/Remote/ServiceHub/Services/CodeAnalysisService_NavigateTo.cs]: This is a private function that is referenced just from the two functions above that return an IList.

**Comment by pentp** [comment]
I originally encountered these async state machines while developing a code fix and saw these long async callstacks while debugging a `CodeAction`. At least these look like they are invoked frequently.

**Comment by jasonmalinowski** [comment]
> When removing async, validating the change to exception behavior is challenging. This increases the risk associated with the change.

For this particular concern, consumers pretty much always have to treat a synchronous throw and failed task as the same thing. Consider an async function that does an await early on, and might throw an exception after the await. Depending on whether that await completes synchronously, you could get either an exception or failed task....and that's normal.

**Comment by jasonmalinowski** [comment]
@pentp Apologies, it looks like we lost this pull request in the stack. We're still happy to merge it, but we'll need to resolve the merge conflict. Is that something you're happy to do, or would you like us to? If you want us to do it we'll be pushing to your branch likely, so wanted to give you a heads up.

**Comment by jasonmalinowski** [comment]
@dotnet-bot retest this please.

**Comment by pentp** [comment]
I haven't got time to look at the merge conflicts, nothing against pushing to my branch.

**Comment by jasonmalinowski** [comment]
@pentp: good, because I pushed earlier today!

**Comment by jasonmalinowski** [comment]
@jinujoseph: ask mode approval? community PR with small refactorings to improve some perf.

**Comment by jinujoseph** [comment]
Thanks @pentp  , Approved for 15.8.Preview3

**Comment by jasonmalinowski** [comment]
Thank you @pentp for the contribution! Apologies it took so long to get in.
